firedrake/deflation.py

```diff
@@ -0,0 +1,217 @@
+        tau = self.compute_tau(deflation, self.y, dy_pet, vi_inact)
+        dy_pet.scale(tau)
+
+        ksp.setConvergedReason(self.ksp.getConvergedReason())
+
+    def compute_tau(self, deflation, state, update_p, vi_inact):
+        if deflation is not None:
+            Edy = self.getEdy(deflation, state, update_p, vi_inact)
+
+            minv = 1.0 / deflation.evaluate(state)
+            tau = 1/(1 - minv*Edy)
+            return tau
+        else:
+            return 1
+
+    def getEdy(self, deflation, y, dy, vi_inact):
+
+        with deflation.deriv(y).dat.vec as deriv:
+            if vi_inact is not None:
+                deriv_ = deriv.getSubVector(vi_inact)
+            else:
+                deriv_ = deriv
+
+            out = -deriv_.dot(dy)
+
+            if vi_inact is not None:
+                deriv.restoreSubVector(vi_inact, deriv_)
+
+        return out
+
+    def reset(self, ksp):
+        self.ksp.reset()
+
+    def view(self, ksp, viewer):
+        self.ksp.view(viewer)
+
+
+class Deflation:
+    """
+    The shifted deflation operator presented in doi:10.1137/140984798.
+    Defaults to power 2, shift 1, and the L2 norm for distance.
+    """
+    def __init__(self, roots=None, power=2, shift=1, op=None):
+        self.power = power
+        self.shift = shift
+        self.roots = list(roots if roots else [])
+
+        if op is None:
+            op = lambda x, y: inner(x - y, x - y)*dx
+        self.op = op
+
+        self.append = self.roots.append
+
+    def evaluate(self, y):
+        """Evaluate the value of the deflation operator, at the current guess y."""
+        from firedrake import assemble
+
+        m = 1.0
+        for root in self.roots:
+            normsq = assemble(self.op(y, root))
+            factor = normsq**(-self.power/2.0) + self.shift
+            m *= factor
+
+        return m
+
+    def deriv(self, y):
+        """Evaluate the derivative of the deflation operator, at the current guess y."""
+        from firedrake import Cofunction, assemble, derivative
+        from ufl import product
+
+        if len(self.roots) == 0:
+            deta = Cofunction(y.function_space().dual())
+            return deta
+
+        p = self.power
+        factors = []
+        dfactors = []
+        dnormsqs = []
+        normsqs = []
+
+        for root in self.roots:
+            form = self.op(y, root)
+            normsqs.append(assemble(form))
+            dnormsqs.append(assemble(derivative(form, y)))
+
+        for normsq in normsqs:
+            factor = normsq**(-p/2.0) + self.shift
+            dfactor = (-p/2.0) * normsq**((-p/2.0) - 1.0)
+
+            factors.append(factor)
+            dfactors.append(dfactor)
+
+        eta = product(factors)
+
+        deta = Cofunction(y.function_space().dual())
+
+        for (_, factor, dfactor, dnormsq) in zip(self.roots, factors, dfactors, dnormsqs):
+            deta.assign(float((eta/factor)*dfactor)*dnormsq + deta)
```

Comment:
This is moslty style, and to avoid explicit calls to `Cofunction` and reference to a dual space. Using float here might be okay even in complex mode, but I don't see it necessary.
We might need `from firedrake.assemble import assemble` at the top of the function.
```suggestion
        deta = assemble(sum(((eta/factor)*dfactor) * dnormsq
                            for (_, factor, dfactor, dnormsq) in zip(self.roots, factors, dfactors, dnormsqs)))
```

Reply:
Why is `self.roots` there if it not being used?